harvester/DBInterface.py

```diff
@@ -267,41 +225,171 @@ def delete_record(self, record):
-				cur.execute(self._prep("DELETE from descriptions where record_id=?"), (record['record_id'],))
-				cur.execute(self._prep("DELETE from tags where record_id=?"), (record['record_id'],))
-				cur.execute(self._prep("DELETE from geospatial where record_id=?"), (record['record_id'],))
-				cur.execute(self._prep("DELETE from domain_metadata where record_id=?"), (record['record_id'],))
-				cur.execute(self._prep("DELETE from access where record_id=?"), (record['record_id'],))
+				self.delete_related_records("records_x_access", record['record_id'])
+				self.delete_related_records("records_x_creators", record['record_id'])
+				self.delete_related_records("records_x_publishers", record['record_id'])
+				self.delete_related_records("records_x_rights", record['record_id'])
+				self.delete_related_records("records_x_subjects", record['record_id'])
+				self.delete_related_records("records_x_tags", record['record_id'])
+				self.delete_related_records("descriptions", record['record_id'])
+				self.delete_related_records("geospatial", record['record_id'])
+				self.delete_related_records("domain_metadata", record['record_id'])
 			except:
 				self.logger.error("Unable to delete related table rows for record %s" % (record['local_identifier'],))
 				return False
 
 		self.logger.debug("Marked as deleted: record %s" % (record['local_identifier'],))
 		return True
 
-	def get_record_id(self, repo_id, local_identifier):
-		returnvalue = None
+	def delete_related_records(self, crosstable, record_id):
 		con = self.getConnection()
 		with con:
 			cur = self.getCursor(con)
-			cur.execute(self._prep("select record_id from records where local_identifier=? and repository_id = ?"),
-						(local_identifier, repo_id))
+			try:
+				sqlstring1 = "DELETE from {} where record_id=?".format(crosstable)
+				cur.execute(self._prep(sqlstring1), (record_id,))
+			except:
+				return False
+
+		return True
+
+	def get_table_id_column(self, tablename):
+		if tablename == "rights":
+			return "rights_id"
+		elif tablename == "access":
+			return "access_id"
+		elif tablename == "repositories":
+			return "repository_id"
+		elif tablename == "geospatial":
+			return "geospatial_id"
+		elif tablename == "domain_metadata":
+			return "metadata_id"
+		elif tablename == "domain_schemas":
+			return "schema_id"
+		elif "records_x_" in tablename:
+			return tablename + "_id"
+		else:
+			return tablename[:-1] + "_id"
+
+	def get_table_value_column(self, tablename):
+		if tablename == "rights":
+			return "rights_hash"
+		elif tablename == "access":
+			return "access"
+		elif tablename == "geospatial":
+			return "coordinate_type"
+		elif tablename == "domain_metadata":
+			return "schema_id"
+		elif tablename == "domain_schemas":
+			return "namespace"
+		elif tablename == "repositories":
+			return "repository_url"
+		elif tablename == "records":
+			return "local_identifier"
+		else:
+			return tablename[:-1]
```

Comment:
This is a little messy but I get what it's doing, might be better to have k:v pairs defined somewhere else if we have to differentiate "the pkey" and "the value" from the various tables more often. The `tablename[:-1]` is elegant but it seems somewhat incidental to how we've done things so far so I'm afraid it might be brittle when adding new tables if it's not more obvious elsewhere in the code...

Reply:
Particularly since the `insert_related_record` function now accepts kwargs (which is also a good change)